DB Query Monitor

felipeelia opened the pull request:
### Description of the Change

This PR adds a "submodule" to Support Monitor, logging, and sending heavy SQL queries.

It also remove some `instance()->setup()` calls, as `instance()` already calls `setup()` in its first execution (what was causing hooks to be added twice.)

### Benefits

While updating plugins on staging we can detect any major changes in the database structure, so we can plan accordingly the same changes for production.

### Verification Process

1. Installed the plugin in a WordPress test installation
2. Installed Support Monitor in another WP test install
3. Checked messages in the "system" group coming

### Changelog Entry

Added: DB Query Monitor

bengreeley commented on includes/classes/SupportMonitor/DBQueryMonitor.php:
```diff
@@ -0,0 +1,173 @@
+namespace TenUpExperience\SupportMonitor;
+
+/**
+ * DBQueryMonitor class
```
@felipeelia I'd love to see more comments describing what this class does and how it works in plain terms so there isn't any concern from installing this and assumptions of anything negative that may be happening with the queries. For example, I'd like to see a description of what the class checks for, that it only uses non-production environments, etc.

felipeelia replied:
@bengreeley I put part of that in the doc description (`A submodule of Support Monitor to report heavy SQL queries executed on staging.`). As we are still defining what is a "heavy SQL query", I'll revisit it later but you made me curious about the best location of that. Is it in the class doc block or in the file's one?

bengreeley replied:
I think this is good for now! I'm sure at some point we can improve documentation in general. Thank you

bengreeley commented on includes/classes/SupportMonitor/DBQueryMonitor.php:
```diff
@@ -0,0 +1,173 @@
+			return $query;
+		}
+
+		if ( apply_filters( 'tenup_experience_log_query', true, $query ) ) {
```
@felipeelia Minor, but we should check for this as the first line in this function (checking for false) since if we don't want to log queries, we might as well not bother with the other checks and can bail early on.

felipeelia replied:
@bengreeley if people don't want to log queries at all, they should use the `tenup_experience_log_heavy_queries` filter, so the plugin doesn't even enter in the method. The intent of this `tenup_experience_log_query` is just to give the dev a way to refine things even further ("I don't want to log any queries related to the X plugin", for example). That said, if we move this filter to the beginning of the method, you'd prefer having it logging queries if the filter value is true *and* if it passes the other checks or *or*?

bengreeley replied:
I see. In that case, some additional docblocks about what the filters are intended to be used for would be helpful. My assumption was that this would be to toggle whether or not the logging functionality was activated at all.

bengreeley commented on includes/classes/SupportMonitor/Monitor.php:
```diff
@@ -461,6 +471,7 @@ public function send_daily_report() {
 			$this->format_message(
 				[
 					'php_version' => $this->get_php_version(),
+					'db_queries'  => $this->db_query_monitor->get_report(),
```
@felipeelia Do you know what size the data we can expect with 'db_queries'? My only concern is sending so much data that it pushes the limits of what can be sent through a message (and subsequently stored on the other database). I'd love to see some comments about expectations of what is sent and how we can make sure we're limiting it to a sane amount of data in the case there were 1000 alter statements called with a runaway script.

felipeelia replied:
@bengreeley the size of the data is a decision that still needs to be made. The 1,000 alter statements are somewhat related to the counter you asked in the other comment, by the way. If that is the same query, it would just have the count increased.

felipeelia replied:
With the new code, it is simply sending if heavy db queries were detected or not. The queries are only available on the site, not in Support Monitor anymore.

bengreeley commented on includes/classes/SupportMonitor/DBQueryMonitor.php:
```diff
@@ -0,0 +1,173 @@
+	 * @return array
+	 */
+	public function get_report() {
+		$date_limit = strtotime( '7 days ago' );
```
@felipeelia Messages are sent twice/day. My concern with only removing data within the past 7 days is we're going to have a lot of duplicate information sent during that 7-day period. It might make sense to remove the data as soon as the report data has been generated so we're only working with fresh data.

felipeelia replied:
@bengreeley the scenario I had in mind was: a Sys Eng made a plugin upgrade on staging on Friday but the Web Eng responsible for checking if queries are a concern or not are only able to check it later, on Monday, for example. As messages are erased, wouldn't we lose those queries?

bengreeley replied:
The way Support Monitor works is the 10up Experience plugin sends messages to the Support Monitor dashboard 2x daily, which is then turned into the latest report. In the scenario you mention as far as upgrading on Friday but checking for queries on Monday, you're right -- this wouldn't really catch those queries because they would have been logged already. You might want to consider the retention and see what makes sense. If there's an ALTER statement and it shows up for 7 days, is that accomplishing anything, or would it make more sense to be able to view that information directly on the site that is storing the data and just get an alert in Support Monitor that you should view the log because an excessive ALTER statement was called.

Support Monitor's main purpose is to keep an eye on things that may be problematic - not necessarily to store a bunch of data that gets wiped out. I'd recommend to reframe the approach of sending a message to Support Monitor of the status (like a count of ALTERs or other items in the log) and then have a place in the 10up Experience plugin you could go to in the WP admin to view the log, since those could be quite large.

felipeelia replied:
It is now storing the queries on the site, so the only thing in Support Monitor is if heavy queries were detected or not.

bengreeley commented on includes/classes/SupportMonitor/DBQueryMonitor.php:
```diff
@@ -0,0 +1,173 @@
+			$stored_queries[ $current_date ][ $key ]['count']++;
+		} else {
+			$stored_queries[ $current_date ][ $key ] = [
+				'query' => stripslashes( $query ),
```
@felipeelia How will we be making sure $query doesn't contain any sensitive information such as user e-mail addresses or other information? We'll want to be really careful about sending that sort of sensitive information to Support Monitor where we could be potentially liable for storing that information if we were ever hacked. I'd recommend we do some sort of cleanup or see if you can get the version of the query that has placeholders like %s in place of actual data.

felipeelia replied:
@bengreeley my bad on this one. Trying to get the version of the query with placeholders can be hard (or near impossible depending on the plugin), so cleaning it up would be probably easier. I'll work on that, thanks!

felipeelia replied:
@bengreeley with the change I made, I'm curious to know your opinion about this topic. We are not sending the query anymore but only storing it in a transient. As the entire feature is only available in non-production environments, do you think we still need to escape the queries? (I don't want to make it insecure but also don't want to over-engineer it). If we decide to escape them, would you be okay with using a lib like [this](https://github.com/greenlion/PHP-SQL-Parser)?

bengreeley replied:
I wonder if you create a regex or search/replace to remove anything between '' and "" if that would be sufficient? I'm not as worried if we're storing it on the server, but we'll want to be aware that if we have queries that have sensitive user information that's available to anybody or any plugin, we'll want to think of a way to clean it up.As long as it's information that is freely available in any other WP table it shouldn't be a bit deal, but should be given adequate testing.

felipeelia replied:
@bengreeley I've pushed a commit that does that. When I save a long post these are the queries being saved (the post update and the revision created):

```
Query: UPDATE `wp_posts` SET `post_author` = 1, `post_date` = ?, `post_date_gmt` = ?, `post_content` = ?, `post_content_filtered` = ?, `post_title` = ?, `post_excerpt` = ?, `post_status` = ?, `post_type` = ?, `comment_status` = ?, `ping_status` = ?, `post_password` = ?, `post_name` = ?, `to_ping` = ?, `pinged` = ?, `post_modified` = ?, `post_modified_gmt` = ?, `post_parent` = 0, `menu_order` = 0, `post_mime_type` = ?, `guid` = ? WHERE `ID` = 7
File: /var/www/html/wp-includes/post.php
Line: 4196
Count: 1

Query: INSERT INTO `wp_posts` (`post_author`, `post_date`, `post_date_gmt`, `post_content`, `post_content_filtered`, `post_title`, `post_excerpt`, `post_status`, `post_type`, `comment_status`, `ping_status`, `post_password`, `post_name`, `to_ping`, `pinged`, `post_modified`, `post_modified_gmt`, `post_parent`, `menu_order`, `post_mime_type`, `guid`) VALUES (1, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, 7, 0, ?, ?)
File: /var/www/html/wp-includes/post.php
Line: 4219
Count: 1
```

Does that look okay?

bengreeley commented on includes/classes/SupportMonitor/DBQueryMonitor.php:
```diff
@@ -0,0 +1,173 @@
+	 *            'query' => 'ALTER TABLE wp_my_db_heavy_plugin CHANGE COLUMN `id` id BIGINT(20) UNSIGNED NOT NULL AUTO_INCREMENT',
+	 *            'file' => '/var/www/html/wp-content/plugins/my-db-heavy-plugin/my-db-heavy-plugin.php',
+	 *            'line' => 53,
+	 *            'count' => 1,
```
@felipeelia Curious what the 'count' attribute will be doing. It seems to me we'd be able to get this via the order the data is stored. Not a big deal, but I wanted to point it out.

felipeelia replied:
@bengreeley my intent with it was just to avoid duplicate entries on the same day. So, for example, we had a plugin that tried to alter the same table several times. Instead of storing the same queries, the code would be just increasing that count. Makes sense?

bengreeley replied:
Yep makes sense to me

bengreeley commented on includes/classes/SupportMonitor/DBQueryMonitor.php:
```diff
@@ -0,0 +1,173 @@
+	public function maybe_log_query( $query ) {
+		global $wpdb;
+
+		if ( ! preg_match( '/^\s*(create|alter|truncate|drop|insert|delete|update|replace)\s/i', $query ) ) {
```
@felipeelia I'm not sure if this is intentional, but I want to point out that logging DELETE or UPDATE statements are likely going to fill the database if you figure in the case we are running a migration or just some heavy maintenance is being done on a non-production environment (like preprod). I also question the value of having that information logged since it's so common. Maybe there's a way to identify whether it's not a normal WP action that was initiated from the backend?

felipeelia replied:
@bengreeley yeah, that is still being worked out. The plan right now is to log only insert/update/delete queries with a nested select or queries that are really big. I'll work on it during the next few days. Thanks!

felipeelia replied:
this has been added to the code already ^

bengreeley reviewed (changes requested):
@felipeelia I think this is a good start - I left some comments and suggestions, let me know if you have any questions

felipeelia commented:
@bengreeley I've refactored the code a bit to display the queries in the site rather than sending them to Support Monitor and also added some docs. Do you mind giving it another look? Thank you very much for all the feedback so far!

bengreeley commented:
> @bengreeley I've refactored the code a bit to display the queries in the site rather than sending them to Support Monitor and also added some docs. Do you mind giving it another look? Thank you very much for all the feedback so far!
> 
> ps.: I'm still considering this WIP as I have to test it with a multisite and also add a slack message in the case any heavy query was detected. Any thoughts on this, btw?

@felipeelia To integrate with Slack, you'll want to look in the Support Monitor Dashboard repo and there will be checks performed during the message ingestion. I'd recommend to send this information with the 'systems' insight so you can piggy-back off that functionality and then in the message ingestion code you would want to check for the heavy query variable that was sent via the message.

felipeelia commented:
@bengreeley I've refactored the code to make the entire feature opt-in, rather than opt-out. You'll see that there are two core concepts: availability (the feature is only available in non-prod envs by default) and state (enabled/disabled). To have it working, it needs to be available AND enabled.

For the slack integration, I talked to @tott and we are seeing it as a "phase two", so let's ignore that for now.

Do you mind giving it yet another review and sharing your thoughts about it? When it looks okay, I need to change the `@since x.x` strings to the actual version.

Thanks!

bengreeley commented:
I think the changes look better, thanks for making those @felipeelia ! We'll want to do thorough testing on some internal environments before using this on production environments and also test with Support Monitor to make sure that data is brought over correctly. You'll likely need to make some changes to that code to accommodate for this new field/data. There's a staging environment we can test on as well if we want to test these changes on a staging version of Support Monitor.

 I'll hold off on approving this until it's been tested, but so far so good!

felipeelia commented:
Thanks, @bengreeley! I'll reach out internally about the best environments to use to test it. Regarding changes needed in the Support Monitor, I'm already testing it locally and as the field is just added to the message, it worked out of the box :)
